Add DateTime input type

ShaeGriffiths opened the pull request:
This has the introduction of the DateTime Input type. This will handle DateOnly, TimeOnly and DateTime.
There is also an introduction of an InputFormat attribute, which allows for the override of the Input display type. This is currently only used in Date or Time inputs.

aakison commented on ExtraDry/ExtraDry.Blazor/Components/Dry/Forms/DryInputDateTime.razor:
```diff
@@ -0,0 +1,4 @@
+﻿@namespace ExtraDry.Blazor.Forms
+@typeparam T
+
+<input class="@CssClasses" title="@Property?.FieldCaption" type="@DisplayMode" value="@Value" readonly="@ReadOnly" @onchange="HandleChange" @attributes="@UnmatchedAttributes" />
```
One of the problems with styling that I've gone around with David is that of the affordance.  The calendar doesn't look too bad on Chrome, but when using input type=search, their magnifying glass is horrible. Given the discussion around currency, should we go ahead and wrap this in a div with an icon and an affordance?  Then, hide the browser calendar affordance like you did with the up/down arrows?

ShaeGriffiths replied:
Given that there's going to be additional work around a new attribute to pass the icon and affordance through which is needed for the numeric input, do you want me to merge the numeric work into this PR for one super-PR or add the affordance and icon for this input into that piece of work.

aakison commented on ExtraDry/ExtraDry.Blazor/Components/Dry/Forms/DryInputDateTime.razor.cs:
```diff
@@ -0,0 +1,113 @@
+            return;
+        }
+        var value = args.Value;
+        if(Property.PropertyType == typeof(DateTime)) {
```
Swallowing a lot of formatting issues without feedback.  Does this run the risk of having the underlying value not the same as the displayed value?

ShaeGriffiths replied:
It does, I just edited the html to allow me to make the input invalid, I'm trying to find a precedent for validation messages from inputs to follow in ED and failing that, i'll see what microsoft have to say on the matter

ShaeGriffiths replied:
DryInput has a valid property and validation message which are displayed, but none of the controls that it renders have a way to populate this, I'll look at adding something to pass in that can be fed back.

ShaeGriffiths replied:
I've passed in a function that can leverage the validation mechanism that already exists on the DryInput

aakison commented on ExtraDry/ExtraDry.Blazor/Models/PropertyDescription.cs:
```diff
@@ -43,6 +43,7 @@ public PropertyDescription(PropertyInfo property)
         }
         --recursionDepth;
         PropertyType = Property.PropertyType.IsGenericType && Property.PropertyType.GetGenericTypeDefinition() == typeof(Nullable<>) ? Property.PropertyType.GetGenericArguments()[0] : Property.PropertyType;
+        InputType = Property.GetCustomAttribute<InputFormatAttribute>()?.DataTypeOverride ?? PropertyType;
```
Mismatch between names "InputType" and "InputFormat", can it just be called InputFormat?

ShaeGriffiths replied:
It can be. This is the InputFormats override type or the actual type of the entity, so it's the Type of the control to render. Happy to align the names, my thought on not aligning is that the InputFormat Attribute may eventually have multiple properties to change the input and didn't want confusion

ShaeGriffiths replied:
I've renamed to match

aakison commented on ExtraDry/ExtraDry.Core/Core/InputFormatAttribute.cs:
```diff
@@ -0,0 +1,8 @@
+namespace ExtraDry.Core;
+
+[AttributeUsage(AttributeTargets.Property, AllowMultiple = false)]
+public class InputFormatAttribute : Attribute
```
Docs...

ShaeGriffiths replied:
Summary tags added

ShaeGriffiths commented:
Because of mismatching holidays/days off, I've gone and added the numeric input changes in here as well. This brings in a few other changes
- Both numeric and date type fields have an icon and affordance.
- I've updated the sample to have these work and look passable (image below - with one field with focus, and font awesome icons being used)
- Number input is actually a text input, but will mutate the data as it is input to remove non-numeric characters
- Binding on the numeric input works a little differently with the binding to allow for this mutation, but it's future-safe for basic calculations

![image](https://github.com/user-attachments/assets/6b304517-e33b-4a6c-8bf4-c99134450433)
